File: python/boggleSolver.py
import trie
import fileToTrie

class BoggleSolver:
	"""find all possible words formed in a grid of chars
		within the given dicationary. Handles adjacent movements 
		between grids including diagonal movements. 8 possible movements from 
		one position on the grid. No repeats
	"""

	def __init__(self, dictionaryText):
		# the file name of the dictionary is passed in
		self.dictionary=fileToTrie.fileToTrie(dictionaryText)


	def solveGrid(self, grid):
		#produce a set of words given a valid grid. A valid grid is any rectangluar grid.

		if self.isValidGrid(grid):
			self.numRow=len(grid)
			self.numCol=len(grid[0])
			self.boogleGrid=grid
			self.setOfWords=[]
			self.visitedGrid=[[False for i in range(self.numCol)] for j in range(self.numRow)]

			for row in range(self.numRow):
				for col in range(self.numCol):
					self.dfs("", row, col)

			self.setOfWords=set(self.setOfWords)
			return self.setOfWords
		else:
			return []


	def isValidGrid(self, grid):
		# checks if it's a valid rectangluar grid
		if len(grid) <= 0:
			return False

		col=len(grid[0])

		for row in range(len(grid)):
			if len(grid[row]) != col:
				return False

		return True



	def dfs(self, word, row, col):
		#checks all possible combination of valid words in the grid

		if row < 0  or row >= self.numRow or col < 0 or  col >= self.numCol or self.visitedGrid[row][col]:
			return
		
		newWord=word+self.boogleGrid[row][col]

		if self.dictionary.isWord(newWord):
			self.setOfWords.append(newWord)

		if self.dictionary.isPrefix(newWord):
			self.visitedGrid[row][col]=True

			for i in range(-1,2):
				for j in range(-1,2):
					self.dfs(newWord, row+i, col+j)
					
			self.visitedGrid[row][col]=False

		return

File: Nodejs/boggleSolver.js
function BoggleSolver(dictionaryFile){
	this.trie=require("./trieDictionary.js").dictionaryToTrie(dictionaryFile);
	this.visited=[];
	this.grid=[];
	this.sets=require("./set");
	this.setOfWords=new this.sets.Set();
}

BoggleSolver.prototype.solveGrid = function(gridToSolve){
	if(!this.isValidGrid(gridToSolve)){
		return [];
	}

	this.grid = gridToSolve;
	this.visited = this.createVisitedGrid(gridToSolve);
	this.setOfWords = new this.sets.Set();
	this.solveGridOnAllPositions();
	var arrayOfWords = this.setOfWords.toArray();
	return arrayOfWords;
}

BoggleSolver.prototype.isValidGrid = function(gridToValidate){
	if(gridToValidate instanceof Array && gridToValidate.length > 0){
		if(gridToValidate[0] instanceof Array && gridToValidate[0].length > 0){
			return true;
		}
	}

	return false;
}

BoggleSolver.prototype.createVisitedGrid = function(gridToCreate){
	var visited = [];

	for(var row = 0; row < gridToCreate.length; row++){
		var visitedRow = [];
		for(var col = 0; col < gridToCreate[row].length; col++){
			visitedRow.push(false);
		}
		visited.push(visitedRow);
	}

	return visited
}


BoggleSolver.prototype.solveGridOnAllPositions = function(){
	for(var row = 0; row < this.grid.length; row++){
		for(var col = 0; col < this.grid[row].length; col++){
			this.findWord("", row, col);
		}
	}
}

BoggleSolver.prototype.findWord = function(wordSoFar, row, col){
	if(row < 0 || row >= this.grid.length || col < 0 || col >= this.grid[row].length || this.visited[row][col]){
		return;
	}

	var newWordSoFar = wordSoFar + this.grid[row][col];

	if(this.trie.isWord(newWordSoFar)){
		this.setOfWords.add(newWordSoFar);
	}

	if(this.trie.isPrefix(newWordSoFar)){
		this.visited[row][col] = true;
		for(var i = -1; i < 2; i++){
			for(var j = -1; j < 2; j++){
				this.findWord(newWordSoFar, row + i, col + j);
			}
		}
		this.visited[row][col] = false;
	}
	return;
}

module.exports.BoggleSolver=BoggleSolver;


